File: UsbCtrl.py
import ConfigVal as CV
import SLABHIDtoSMBUS as SS
import time

i2cAddr = CV.I2CSLAVEADDR * 2
instance = SS.HidSmbusDevice()

def UsbGetNumDevices():
    try:
        numDevices = SS.GetNumDevices(CV.VID, CV.PID)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval, numDevices

def UsbGetOpenedSnDevice(index):
    try:
        SnString = instance.GetString(SS.HID_SMBUS.SERIAL_STR)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval, SnString

def UsbGetSnDevice(index):
    try:
        SnString = SS.GetString(index, CV.VID, CV.PID, SS.HID_SMBUS.SERIAL_STR)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval, SnString

def UsbOpen(deviceNumber):
    try:
        instance.Open(deviceNumber, CV.VID, CV.PID)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval

def UsbClose():
    try:
        instance.Close()
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval

def UsbSetConfig():
    try:
        instance.SetSmbusConfig(CV.BITRATE, i2cAddr, CV.AUTOREADRESPOND,
                                CV.WRITETIMEOUT, CV.READTIMEOUT, CV.SCLLOWTIMEOUT, CV.TRANSFARRETRIES)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
        return retval
    try:
        instance.SetGpioConfig(CV.DIRECTION, CV.MODE, CV.SPECIAL, CV.CLKDIV)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    try:
        instance.SetTimeouts(CV.RESPONSETIMEOUT)
        retval = 0
    except SS.HidSmbusError as e:
        retval =e
    return retval

def UsbRead(segmentOffset, receiveSize):
    sendData = segmentOffset.to_bytes(2, 'big')
    try:
        instance.WriteRequest(i2cAddr, sendData)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    try:
        instance.ReadRequest(i2cAddr, receiveSize)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    time.sleep(0.1)  ## important wait
    try:
        instance.ForceReadResponse(receiveSize)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    try:
        dataLen = 0
        countNum = 0
        while(dataLen == 0):
            receivedData = instance.GetReadResponse(i2cAddr, 62)
            dataLen = len(receivedData)
            if ( countNum > CV.RETRYNUM):
                retval = "USB Read Time Out"
                break
            countNum += 1
            retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval, receivedData

def UsbRead2BytesInt(segmentOffset):
    retval, data = UsbRead(segmentOffset, CV.DATA_LENGTH)
    IntData = int.from_bytes(data, 'big')
    return retval, IntData

def CountRead(segmentOffset):
    retval, data = UsbRead(segmentOffset,CV.LENSCOUNT_LENGTH)
    count = int.from_bytes(data, 'big')
    return retval, count    

def UsbWrite(segmentOffset, writeData):
    sendData1 = segmentOffset.to_bytes(2, 'big')
    sendData2 = writeData.to_bytes(2, 'big')
    sendData = sendData1 + sendData2
    try:
        instance.WriteRequest(i2cAddr,sendData)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval

def UsbWriteString(segmentOffset, writeData):
    sendData1 = segmentOffset.to_bytes(2, 'big')
    sendData2 = bytes(writeData, encoding='UTF-8')
    sendData = sendData1 + sendData2
    try:
        instance.WriteRequest(i2cAddr,sendData)
        retval = 0
    except SS.HidSmbusError as e:
        retval = e
    return retval

File: LensConnect_SampleProgram_Linux_C_x86_x64_1.0.0/LensConnect_Controller/C/UsbCtrl.c
#include "LensCtrl.h"

HID_SMBUS_DEVICE connectedDevice;
BYTE i2cAddr = I2CSLAVEADDR * 2;
BYTE receivedData[80];


int UsbGetNumDevices(DWORD *numDevices) {
	int retval = HidSmbus_GetNumDevices(numDevices, VID, PID);
	return retval;
}

int UsbGetSnDevice(uint16_t index, char* SnString) {
	int retval = HidSmbus_GetString(index, VID, PID, SnString, HID_SMBUS_GET_SERIAL_STR);
	return retval;
}

int UsbOpen(DWORD deviceNumber) {
	int retval = HidSmbus_Open(&connectedDevice, deviceNumber, VID, PID);
	return retval;
}

void UsbClose() {
	HidSmbus_Close(connectedDevice);
}

int UsbSetConfig() {
	int retval = HidSmbus_SetSmbusConfig(connectedDevice,
		BITRATE, i2cAddr, AUTOREADRESPOND,
		WRITETIMEOUT, READTIMEOUT,
		SCLLOWTIMEOUT, TRANSFARRETRIES);
	if (retval == HID_SMBUS_SUCCESS)
		return retval;

	retval = HidSmbus_SetGpioConfig(connectedDevice, DIRECTION, MODE, SPECIAL, CLKDIV);
	if (retval == HID_SMBUS_SUCCESS)
		return retval;

	retval = HidSmbus_SetTimeouts(connectedDevice, RESPONSETIMEOUT);
	if (retval == HID_SMBUS_SUCCESS)
		return retval;

	return retval;
}

int UsbRead(uint16_t segmentOffset, uint16_t receiveSize) {	
	uint8_t sendData[SEGMENTOFFSET_LENGTH];
	sendData[0] = segmentOffset >> 8;
	sendData[1] = (uint8_t)segmentOffset;
	BYTE sendSize = sizeof(sendData);
	int retval = HidSmbus_WriteRequest(connectedDevice, i2cAddr, sendData, sendSize);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;

	retval = HidSmbus_ReadRequest(connectedDevice, i2cAddr, receiveSize);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;

	retval = HidSmbus_ForceReadResponse(connectedDevice, receiveSize);
	if (retval != HID_SMBUS_SUCCESS)
		return retval;

	uint8_t receiveData[128];
	HID_SMBUS_S0 status;
	BYTE totalBytesRead = 0;
	BYTE bufferSize = 62;
	BYTE bytesRead = 0;
	do {
		retval = HidSmbus_GetReadResponse(connectedDevice, &status, 
			receiveData + totalBytesRead, bufferSize, &bytesRead);
		if (retval != HID_SMBUS_SUCCESS)
			return retval;

		totalBytesRead += bytesRead;
	} while (totalBytesRead < receiveSize);

	for (int i = 0; i < totalBytesRead; i++) {
		receivedData[i] = receiveData[i];
	}
	return retval;
}

uint16_t UsbRead2Bytes() {
	return ((receivedData[0] << 8) | receivedData[1]);
}

uint32_t CountRead() {
	return ((receivedData[0] << 24) | (receivedData[1] << 16) |
		 (receivedData[2] << 8) | (receivedData[3]));
}

int UsbWrite(uint16_t segmentOffset, uint16_t writeData) {
	uint8_t sendData[4];
	sendData[0] = segmentOffset >> 8;
	sendData[1] = (uint8_t)segmentOffset;
	sendData[2] = writeData >> 8;
	sendData[3] = (uint8_t)writeData;
	BYTE sendSize = sizeof(sendData);
	int retval = HidSmbus_WriteRequest(connectedDevice, i2cAddr, sendData, sendSize);
	return retval;
}

